The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4139478665/jobs/7157089337) [(plain)](https://github.com/rust-lang-ci/rust/commit/9550d95507c1c9af656f32551702ab3e9f1b9d49/checks/11240259640/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_session test:false 0.980
    Checking rustc_attr v0.0.0 (/checkout/compiler/rustc_attr)
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
error[E0407]: method `store_nocache` is not a member of trait `QueryStorage`
   --> compiler/rustc_query_system/src/query/caches.rs:143:5
    |
143 | /     fn store_nocache(&self, value: Self::Value) -> Self::Stored {
145 | |     }
145 | |     }
    | |_____^ not a member of trait `QueryStorage`

error[E0277]: the trait bound `V: std::marker::Copy` is not satisfied
   --> compiler/rustc_query_system/src/query/caches.rs:122:18
    |
122 |     type Cache = SingleCache<V>
    |                  ^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `V`
help: consider further restricting this bound
    |
    |
121 | impl<'tcx, V: 'tcx + std::marker::Copy> CacheSelector<'tcx, V> for SingleCacheSelector {


error[E0277]: the trait bound `V: std::marker::Copy` is not satisfied
   --> compiler/rustc_query_system/src/query/caches.rs:140:19
140 |     type Stored = V;
    |                   ^ the trait `std::marker::Copy` is not implemented for `V`
    |
note: required by a bound in `QueryStorage::Stored`
note: required by a bound in `QueryStorage::Stored`
   --> compiler/rustc_query_system/src/query/caches.rs:24:18
    |
24  |     type Stored: Copy;
    |                  ^^^^ required by this bound in `QueryStorage::Stored`
help: consider further restricting this bound
    |
138 | impl<V: Clone + Debug + std::marker::Copy> QueryStorage for SingleCache<V> {


error[E0049]: method `lookup` has 2 type parameters but its trait declaration has 0 type parameters
   --> compiler/rustc_query_system/src/query/caches.rs:155:15
    |
34  |     fn lookup(&self, key: &Self::Key) -> Option<(Self::Stored, DepNodeIndex)>;
    |              - expected 0 type parameters
...
155 |     fn lookup<R, OnHit>(&self, _key: &(), on_hit: OnHit) -> Result<R, ()>
    |               |
    |               found 2 type parameters

[RUSTC-TIMING] rustc_attr test:true 0.279
```

</details>

--- a/compiler/rustc_middle/src/query/keys.rs
+++ b/compiler/rustc_middle/src/query/keys.rs
@@ -8,7 +8,7 @@ use crate::ty::subst::{GenericArg, SubstsRef};
 use crate::ty::{self, layout::TyAndLayout, Ty, TyCtxt};
 use rustc_hir::def_id::{CrateNum, DefId, LocalDefId, LOCAL_CRATE};
 use rustc_hir::hir_id::{HirId, OwnerId};
-use rustc_query_system::query::{DefaultCacheSelector, VecCacheSelector};
+use rustc_query_system::query::{DefaultCacheSelector, SingleCacheSelector, VecCacheSelector};
 use rustc_span::symbol::{Ident, Symbol};
 use rustc_span::{Span, DUMMY_SP};
 
@@ -45,7 +45,7 @@ pub trait Key: Sized {
 }
 
 impl Key for () {
-    type CacheSelector = DefaultCacheSelector<Self>;
+    type CacheSelector = SingleCacheSelector;
 
     #[inline(always)]
     fn query_crate_is_local(&self) -> bool {

--- a/compiler/rustc_query_system/src/query/caches.rs
+++ b/compiler/rustc_query_system/src/query/caches.rs
@@ -5,7 +5,6 @@ use rustc_data_structures::fx::FxHashMap;
 use rustc_data_structures::sharded;
 #[cfg(parallel_compiler)]
 use rustc_data_structures::sharded::Sharded;
-#[cfg(not(parallel_compiler))]
 use rustc_data_structures::sync::Lock;
 use rustc_data_structures::sync::WorkerLocal;
 use rustc_index::vec::{Idx, IndexVec};
@@ -114,6 +113,52 @@ where
                 f(k, &v.0, v.1);
             }
         }
+    }
+}
+
+pub struct SingleCacheSelector;
+
+impl<'tcx, V: 'tcx> CacheSelector<'tcx, V> for SingleCacheSelector {
+    type Cache = SingleCache<V>
+    where
+        V: Copy;
+    type ArenaCache = ArenaCache<'tcx, (), V>;
+}
+
+pub struct SingleCache<V> {
+    cache: Lock<Option<(V, DepNodeIndex)>>,
+}
+
+impl<V> Default for SingleCache<V> {
+    fn default() -> Self {
+        SingleCache { cache: Lock::new(None) }
+    }
+}
+
+impl<V: Copy + Debug> QueryStorage for SingleCache<V> {
+    type Value = V;
+    type Stored = V;
+}
+
+impl<V> QueryCache for SingleCache<V>
+where
+    V: Copy + Debug,
+{
+    type Key = ();
+
+    #[inline(always)]
+    fn lookup(&self, _key: &()) -> Option<(V, DepNodeIndex)> {
+        *self.cache.lock()
+    }
+
+    #[inline]
+    fn complete(&self, _key: (), value: V, index: DepNodeIndex) -> Self::Stored {
+        *self.cache.lock() = Some((value.clone(), index));
+        value
+    }
+
+    fn iter(&self, f: &mut dyn FnMut(&Self::Key, &Self::Value, DepNodeIndex)) {
+        self.cache.lock().as_ref().map(|value| f(&(), &value.0, value.1));
     }
 }
 

--- a/compiler/rustc_query_system/src/query/mod.rs
+++ b/compiler/rustc_query_system/src/query/mod.rs
@@ -8,7 +8,8 @@ pub use self::job::{print_query_stack, QueryInfo, QueryJob, QueryJobId, QueryJob
 
 mod caches;
 pub use self::caches::{
-    CacheSelector, DefaultCacheSelector, QueryCache, QueryStorage, VecCacheSelector,
+    CacheSelector, DefaultCacheSelector, QueryCache, QueryStorage, SingleCacheSelector,
+    VecCacheSelector,
 };
 
 mod config;

